### PR TITLE
STORY-909/908/907: is-true, is-false & condition block

### PR DESCRIPTION
**Pull request by juliobranha**
https://alleyinteractive.atlassian.net/browse/STORY-909
https://alleyinteractive.atlassian.net/browse/STORY-908
https://alleyinteractive.atlassian.net/browse/STORY-907

### CODE REVIEW

File: blocks/condition/block.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	"title": "Condition",
+	"category": "theme",
+	"icon": "star-half",
+	"description": "Conditionally show is-true or is-false blocks",
+	"textdomain": "condition",
+	"editorScript": "file:index.ts",
+	"editorStyle": "file:index.css",
+	"style": [
+		"file:style-index.css"
+	],
```
Comment (goodguyry) [inline blocks/condition/block.json]: This block won't have frontend styles

```suggestion
```

File: blocks/condition/block.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    },
+    "post": {
+      "type": "string"
+    },
+    "query": {
+      "type": "string"
+    },
+    "index": {
+      "type": "string"
+    }
```
Comment (goodguyry) [inline blocks/condition/block.json]: `index` is an array in the project from which these blocks originated.

File: blocks/condition/edit.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            onChange={(next) => setAttributes({ query: next })}
+            value={query}
+          />
+        </PanelRow>
+
+        { /* @ts-ignore */ }
+        <PanelRow>
+          { /* @ts-ignore */ }
+          <TextControl
+            label={__('Index', 'wp')}
```
Comment (goodguyry) [inline blocks/condition/edit.tsx]:
```suggestion
            label={__('Index', 'wp-curate')}
```

File: blocks/condition/edit.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            onChange={(next) => setAttributes({ index: next })}
+            value={index}
+          />
+        </PanelRow>
+
+        { /* @ts-ignore */ }
+        <PanelRow>
+          { /* @ts-ignore */ }
+          <TextControl
+            label={__('Post', 'wp')}
```
Comment (goodguyry) [inline blocks/condition/edit.tsx]:
```suggestion
            label={__('Post', 'wp-curate')}
```

File: blocks/condition/edit.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            onChange={(next) => setAttributes({ post: next })}
+            value={post}
+          />
+        </PanelRow>
+
+        { /* @ts-ignore */ }
+        <PanelRow>
+          { /* @ts-ignore */ }
+          <TextControl
+            label={__('Custom', 'wp')}
```
Comment (goodguyry) [inline blocks/condition/edit.tsx]:
```suggestion
            label={__('Custom', 'wp-curate')}
```

File: blocks/condition/edit.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            onChange={(next) => setAttributes({ custom: next })}
+            value={custom}
+          />
+        </PanelRow>
+
+        { /* @ts-ignore */ }
+        <PanelRow>
+          { /* @ts-ignore */ }
+          <TextControl
+            label={__('Condition', 'wp')}
```
Comment (goodguyry) [inline blocks/condition/edit.tsx]:
```suggestion
            label={__('Condition', 'wp-curate')}
```

File: blocks/condition/index.tsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  edit,
+  save: () => {
+    const blockProps = useBlockProps.save();
+    return (
+      <div {...blockProps}>
+        {/* @ts-ignore */}
+        <InnerBlocks.Content />
+      </div>
+    );
+  },
```
Comment (goodguyry) [inline blocks/condition/index.tsx]: I'm not sure there's much to gain by assigning the `save` function to a variable

```suggestion
  save: () => (
    <div {...useBlockProps.save()}>
      {/* @ts-ignore */}
      <InnerBlocks.Content />
    </div>
  ),
```

File: blocks/condition/style.scss
Comment (goodguyry) [inline blocks/condition/style.scss:1]: I think we can safely remove this file

File: blocks/is-false/block.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+	"title": "Is false",
+	"category": "theme",
+	"icon": "star-empty",
+	"description": "Displays when the condition is false.",
+	"textdomain": "is-false",
+	"editorScript": "file:index.ts",
+	"editorStyle": "file:index.css",
+	"style": [
+		"file:style-index.css"
+	],
```
Comment (goodguyry) [inline blocks/is-false/block.json]: This block won't have frontend styles.

```suggestion
```

File: blocks/is-false/block.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "wp-curate/is-false",
+	"version": "0.1.0",
+	"title": "Is false",
+	"category": "theme",
+	"icon": "star-empty",
+	"description": "Displays when the condition is false.",
+	"textdomain": "is-false",
```
Comment (goodguyry) [inline blocks/is-false/block.json]:
```suggestion
	"textdomain": "wp-curate",
```

File: blocks/condition/block.json [hunk 10]
```diff
@@ -0,0 +1,34 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "wp-curate/condition",
+	"version": "0.1.0",
+	"title": "Condition",
+	"category": "theme",
+	"icon": "star-half",
+	"description": "Conditionally show is-true or is-false blocks",
+	"textdomain": "condition",
```
Comment (goodguyry) [inline blocks/condition/block.json]:
```suggestion
	"textdomain": "wp-curate",
```

File: blocks/is-false/edit.tsx [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+
+/**
+ * The wp-curate/is-true block edit function.
```
Comment (goodguyry) [inline blocks/is-false/edit.tsx]:
```suggestion
 * The wp-curate/is-false block edit function.
```

File: blocks/is-false/index.tsx [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  edit,
+  save: () => {
+    const blockProps = useBlockProps.save();
+    return (
+      <div {...blockProps}>
+        {/* @ts-ignore */}
+        <InnerBlocks.Content />
+      </div>
+    );
+  },
```
Comment (goodguyry) [inline blocks/is-false/index.tsx]: I'm not sure there's much to gain by assigning the `save` function to a variable
```suggestion
  save: () => (
    <div {...useBlockProps.save()}>
      {/* @ts-ignore */}
      <InnerBlocks.Content />
    </div>
  ),
```

File: blocks/is-false/style.scss
Comment (goodguyry) [inline blocks/is-false/style.scss:1]: I think we can safely remove this file

File: blocks/is-true/block.json [hunk 13]
```diff
@@ -0,0 +1,18 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "wp-curate/is-true",
+	"version": "0.1.0",
+	"title": "Is true",
+	"category": "theme",
+	"icon": "star-filled",
+	"description": "Displays when the condition is true.",
+	"textdomain": "is-true",
```
Comment (goodguyry) [inline blocks/is-true/block.json]:
```suggestion
	"textdomain": "wp-curate",
```

File: blocks/is-true/block.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+	"editorScript": "file:index.ts",
+	"editorStyle": "file:index.css",
+	"style": [
+		"file:style-index.css"
+	],
```
Comment (goodguyry) [inline blocks/is-true/block.json]: This block won't have frontend styles
```suggestion
```

File: blocks/is-true/index.tsx [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  edit,
+  save: () => {
+    const blockProps = useBlockProps.save();
+    return (
+      <div {...blockProps}>
+        {/* @ts-ignore */}
+        <InnerBlocks.Content />
+      </div>
+    );
+  },
```
Comment (goodguyry) [inline blocks/is-true/index.tsx]: I'm not sure there's much to gain by assigning the save function to a variable
```suggestion
  save: () => (
    <div {...useBlockProps.save()}>
      {/* @ts-ignore */}
      <InnerBlocks.Content />
    </div>
  ),
```

File: blocks/is-true/index.tsx [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import { registerBlockType } from '@wordpress/blocks';
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+
+import edit from './edit';
+import metadata from './block.json';
+
+import './style.scss';
```
Comment (goodguyry) [inline blocks/is-true/index.tsx]:
```suggestion
```

File: blocks/condition/index.tsx [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import { registerBlockType } from '@wordpress/blocks';
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+
+import edit from './edit';
+import metadata from './block.json';
+
+import './style.scss';
```
Comment (goodguyry) [inline blocks/condition/index.tsx]:
```suggestion
```

File: blocks/is-false/index.tsx [hunk 18]
```diff
@@ -0,0 +1,21 @@
+import { registerBlockType } from '@wordpress/blocks';
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+
+import edit from './edit';
+import metadata from './block.json';
+
+import './style.scss';
```
Comment (goodguyry) [inline blocks/is-false/index.tsx]:
```suggestion
```

File: blocks/is-true/style.scss
Comment (goodguyry) [inline blocks/is-true/style.scss:1]: I think we can safely remove this file

**Review by goodguyry — COMMENTED** [review]
Lots of suggestions for cleanup. I've not approved nor requested changes because the render files are empty. I'd suggest working with @dlh01 and @mogmarsh to get those fleshed out.

**Comment by dlh01** [comment]
> Lots of suggestions for cleanup. I've not approved nor requested changes because the render files are empty. I'd suggest working with @dlh01 and @mogmarsh to get those fleshed out.

These blocks don't have render files. The logic for their `render_callback` functions appears to be in #14.

**Comment by juliobranha** [comment]
> Lots of suggestions for cleanup. I've not approved nor requested changes because the render files are empty. I'd suggest working with @dlh01 and @mogmarsh to get those fleshed out.

I understand the rendering is not part of these tickets

File: blocks/condition/block.json (hunk 2 above)
Comment (juliobranha) [reply blocks/condition/block.json]: @goodguyry I noticed that but not sure what's best here, I see a few options:

1. Repeater field. Ideal solution but maybe an overkill?
2. Leave the attribute as `string` and convert it to array in PHP. The User should enter comma-separated indexes
3. Keep the text field and convert to array in js, so the attribute gets saved as an `array`. The User should enter comma-separated indexes
4. Allow a max of two indexes, and have two separated fields

Maybe number 3? Let me know what you think is best :) Maybe @dlh01 or @mogmarsh have a preference?

**Review by goodguyry — APPROVED** [review]
🎸 Okay, I see now. There's a lot of overlap between a few PRs here.

**Comment by juliobranha** [comment]
> 🎸 Okay, I see now. There's a lot of overlap between a few PRs here.

Yes, I wasn't aware of the work on the other tickets, maybe these ones are a bit redundant

File: blocks/condition/block.json (hunk 2 above)
Comment (dlh01) [reply blocks/condition/block.json]: For JSON purposes, `index` in that project is an object, not an array, e.g. `{"===": 0}`. There was no use case for needing multiple indices on the project and so how multiple indices should be specified in the schema is uncertain. Unless we're aware of a definite use case now, I'm not sure we need to account for it yet.

Based on the actual uses of `index` in the original project, we need a single set of fields: one to choose the operator and one to select an index. There were no cases of using multiple operators in a condition. If someone really needed to, they could achieve using multiple operators with a new condition block inside an `is-false` block.

Comment (juliobranha) [reply blocks/condition/block.json]: > For JSON purposes, `index` in that project is an object, not an array, e.g. `{"===": 0}`. There was no use case for needing multiple indices on the project and so how multiple indices should be specified in the schema is uncertain. Unless we're aware of a definite use case now, I'm not sure we need to account for it yet.
> 
> Based on the actual uses of `index` in the original project, we need a single set of fields: one to choose the operator and one to select an index. There were no cases of using multiple operators in a condition. If someone really needed to, they could achieve using multiple operators with a new condition block inside an `is-false` block.

Thanks @dlh01, updated

**Pull request merged**